File: src/pvz_console/tools/trello_export.py
from __future__ import annotations

import json
import os
from dataclasses import dataclass, field
from pathlib import Path
from typing import Dict, List, Sequence

from pvz_console.config import PROJECT_ROOT, SOURCE_LOCALE
from pvz_console.core.diff import missing_by_id
from pvz_console.core.models import AlmanacEntry, StringEntry, TrelloCard
from pvz_console.parsers.almanac import load_achievements, load_plants, load_zombies
from pvz_console.parsers.strings import (
    diff_abyss_buffs,
    diff_regexs,
    diff_strings,
    diff_tips_fs,
    diff_tips_iz,
    diff_travel_buffs,
)
from pvz_console.reporting.trello_csv import build_trello_readme, write_trello_csvs_by_list

NAME_MAX_LEN = 100
DESCRIPTION_MAX_LEN = 15000  # Trello hard limit is 16384; leave headroom.
DEFAULT_LABEL = "To be translated"

LIST_PLANTS = "Plants"
LIST_ZOMBIES = "Zombies"
LIST_ACHIEVEMENTS = "Achievements"
LIST_STRINGS = "Strings"
LIST_REGEX = "Regex"
LIST_TIPS_IZ = "Tips IZ"
LIST_TIPS_FS = "Tips FS"
LIST_ABYSS = "Abyss Buffs"
LIST_TRAVEL = "Travel Buffs"


@dataclass
class TrelloExportResult:
    locale: str
    output_dir: str
    csv_paths: Dict[str, str] = field(default_factory=dict)
    readme_path: str = ""
    counts_by_list: Dict[str, int] = field(default_factory=dict)

    @property
    def total_cards(self) -> int:
        return sum(self.counts_by_list.values())


def _truncate(text: str, limit: int) -> str:
    if len(text) <= limit:
        return text
    return text[: limit - 1].rstrip() + "\u2026"


def _fence(body: str) -> str:
    return "```json\n" + body + "\n```"


def _describe_almanac(entry: AlmanacEntry) -> str:
    """Wrap the raw dict (with braces) in a ```json fence \u2014 same shape as the
    source file. Trello renders the fence as a monospace code block with
    preserved line breaks, and \\n stays literal thanks to ``json.dumps``.
    """
    return _fence(json.dumps(entry.raw, ensure_ascii=False, indent=2))


def _describe_string(entry: StringEntry) -> str:
    """Render a flat key/value entry as a JSON member inside a ```json fence:
    ``"key": "value"``. Matches how each line would appear in the source
    ``.json`` file.
    """
    key = json.dumps(entry.key, ensure_ascii=False)
    value = json.dumps(entry.source if entry.source is not None else "", ensure_ascii=False)
    return _fence(f"{key}: {value}")


def _almanac_cards(entries: Sequence[AlmanacEntry], list_name: str, label: str) -> List[TrelloCard]:
    cards: List[TrelloCard] = []
    for entry in entries:
        name = entry.name or f"id {entry.id}"
        cards.append(TrelloCard(
            name=_truncate(name, NAME_MAX_LEN),
            description=_truncate(_describe_almanac(entry), DESCRIPTION_MAX_LEN),
            list_name=list_name,
            labels=label,
        ))
    return cards


def _string_cards(entries: Sequence[StringEntry], list_name: str, label: str) -> List[TrelloCard]:
    cards: List[TrelloCard] = []
    for entry in entries:
        cards.append(TrelloCard(
            name=_truncate(entry.key, NAME_MAX_LEN),
            description=_truncate(_describe_string(entry), DESCRIPTION_MAX_LEN),
            list_name=list_name,
            labels=label,
        ))
    return cards


def collect_cards(root: str, locale: str, label: str = DEFAULT_LABEL) -> List[TrelloCard]:
    cards: List[TrelloCard] = []

    cards += _almanac_cards(
        missing_by_id(load_plants(root, SOURCE_LOCALE), load_plants(root, locale)),
        LIST_PLANTS, label,
    )
    cards += _almanac_cards(
        missing_by_id(load_zombies(root, SOURCE_LOCALE), load_zombies(root, locale)),
        LIST_ZOMBIES, label,
    )
    cards += _almanac_cards(
        missing_by_id(load_achievements(root, SOURCE_LOCALE), load_achievements(root, locale)),
        LIST_ACHIEVEMENTS, label,
    )

    cards += _string_cards(diff_strings(root, SOURCE_LOCALE, locale), LIST_STRINGS, label)
    cards += _string_cards(diff_regexs(root, SOURCE_LOCALE, locale), LIST_REGEX, label)
    cards += _string_cards(diff_tips_iz(root, SOURCE_LOCALE, locale), LIST_TIPS_IZ, label)
    cards += _string_cards(diff_tips_fs(root, SOURCE_LOCALE, locale), LIST_TIPS_FS, label)
    cards += _string_cards(diff_abyss_buffs(root, SOURCE_LOCALE, locale), LIST_ABYSS, label)
    cards += _string_cards(diff_travel_buffs(root, SOURCE_LOCALE, locale), LIST_TRAVEL, label)

    return cards


def export_trello(
    locale: str,
    exports_root: os.PathLike | str,
    root: os.PathLike | str = PROJECT_ROOT,
    label: str = DEFAULT_LABEL,
) -> TrelloExportResult:
    root_str = str(root)
    out = Path(exports_root) / locale
    out.mkdir(parents=True, exist_ok=True)

    cards = collect_cards(root_str, locale, label=label)
    counts: Dict[str, int] = {}
    for card in cards:
        counts[card.list_name] = counts.get(card.list_name, 0) + 1

    csv_paths = write_trello_csvs_by_list(cards, str(out))
    readme_path = out / "trello_README.md"
    build_trello_readme(locale, csv_paths, label, str(readme_path))

    return TrelloExportResult(
        locale=locale,
        output_dir=str(out),
        csv_paths=csv_paths,
        readme_path=str(readme_path),
        counts_by_list=counts,
    )

File: src/tools/trello-export.ts
import { mkdirSync } from "node:fs";
import path from "node:path";

import { PROJECT_ROOT, SOURCE_LOCALE } from "../config";
import { missingById } from "../core/diff";
import {
  DEFAULT_TRELLO_LABEL,
  type AlmanacEntry,
  type StringEntry,
  type TravelBuffEntry,
  type TrelloCard,
} from "../core/models";
import { stripTrailingWhitespace } from "../core/text";
import { loadAchievements, loadPlants, loadZombies } from "../parsers/almanac";
import {
  diffAbyssBuffs,
  diffRegexs,
  diffStrings,
  diffTipsFs,
  diffTipsIz,
  diffTravelBuffs,
} from "../parsers/strings";
import {
  type TrelloListSummary,
  buildTrelloReadme,
  writeTrelloCsvsByList,
} from "../reporting/trello-csv";

export const NAME_MAX_LEN = 100;
export const DESCRIPTION_MAX_LEN = 15000; // Trello hard limit is 16384; leave headroom.

export const LIST_PLANTS = "Plants";
export const LIST_ZOMBIES = "Zombies";
export const LIST_ACHIEVEMENTS = "Achievements";
export const LIST_STRINGS = "Strings";
export const LIST_REGEX = "Regex";
export const LIST_TIPS_IZ = "Tips IZ";
export const LIST_TIPS_FS = "Tips FS";
export const LIST_ABYSS = "Abyss Buffs";
export const LIST_TRAVEL = "Travel Buffs";

const README_FILENAME = "trello_README.md";

/** What to export, and against which reference. */
export interface TrelloExportRequest {
  /** Target locale — the one that still needs translating. */
  locale: string;
  /** `PvZ_Fusion_Translator/` root. Defaults to the discovered project root. */
  root?: string;
  /** Trello label stamped on every card. */
  label?: string;
  /** Reference locale the target is diffed against. */
  source?: string;
}

export interface TrelloExportOptions extends TrelloExportRequest {
  /** Folder receiving `<locale>/`, with the CSVs and the import README. */
  exportsRoot: string;
}

export interface TrelloExportResult {
  locale: string;
  outputDir: string;
  readmePath: string;
  /** One entry per written CSV — empty when the locale is fully translated. */
  lists: TrelloListSummary[];
  totalCards: number;
}

export function truncate(text: string, limit: number): string {
  if (text.length <= limit) return text;
  return `${stripTrailingWhitespace(text.slice(0, limit - 1))}…`;
}

function fence(body: string): string {
  return "```json\n" + body + "\n```";
}

/** Wrap the raw entry (braces included) in a ```json fence — same shape as the source file. */
export function describeAlmanac(entry: AlmanacEntry): string {
  return fence(JSON.stringify(entry.raw, null, 2));
}

/** Render a flat key/value entry as the JSON member `"key": "value"`. */
export function describeString(entry: StringEntry): string {
  return fence(`${JSON.stringify(entry.key)}: ${JSON.stringify(entry.source ?? "")}`);
}

function almanacCards(
  entries: readonly AlmanacEntry[],
  listName: string,
  label: string,
): TrelloCard[] {
  return entries.map((entry) => ({
    name: truncate(entry.name || `id ${entry.id}`, NAME_MAX_LEN),
    description: truncate(describeAlmanac(entry), DESCRIPTION_MAX_LEN),
    listName,
    labels: label,
  }));
}

function stringCards(
  entries: readonly StringEntry[],
  listName: string,
  label: string,
): TrelloCard[] {
  return entries.map((entry) => ({
    name: truncate(entry.key, NAME_MAX_LEN),
    description: truncate(describeString(entry), DESCRIPTION_MAX_LEN),
    listName,
    labels: label,
  }));
}

function travelBuffCards(entries: readonly TravelBuffEntry[], label: string): TrelloCard[] {
  return entries.map((entry) => ({
    name: truncate(entry.source || entry.key, NAME_MAX_LEN),
    description: truncate(
      fence(JSON.stringify({ [entry.category]: { [entry.id]: entry.raw } }, null, 2)),
      DESCRIPTION_MAX_LEN,
    ),
    listName: LIST_TRAVEL,
    labels: label,
  }));
}

/** Resolved form of a {@link TrelloExportRequest}, defaults applied. */
interface ResolvedRequest {
  locale: string;
  root: string;
  label: string;
  source: string;
}

function resolveRequest(request: TrelloExportRequest): ResolvedRequest {
  return {
    locale: request.locale,
    root: request.root ?? PROJECT_ROOT,
    label: request.label ?? DEFAULT_TRELLO_LABEL,
    source: request.source ?? SOURCE_LOCALE,
  };
}

function almanacListCards({ root, locale, label, source }: ResolvedRequest): TrelloCard[] {
  const missingPlants = missingById(loadPlants(root, source), loadPlants(root, locale));
  const missingZombies = missingById(loadZombies(root, source), loadZombies(root, locale));
  const missingAchievements = missingById(
    loadAchievements(root, source),
    loadAchievements(root, locale),
  );
  return [
    ...almanacCards(missingPlants, LIST_PLANTS, label),
    ...almanacCards(missingZombies, LIST_ZOMBIES, label),
    ...almanacCards(missingAchievements, LIST_ACHIEVEMENTS, label),
  ];
}

function stringListCards({ root, locale, label, source }: ResolvedRequest): TrelloCard[] {
  return [
    ...stringCards(diffStrings(root, source, locale), LIST_STRINGS, label),
    ...stringCards(diffRegexs(root, source, locale), LIST_REGEX, label),
    ...stringCards(diffTipsIz(root, source, locale), LIST_TIPS_IZ, label),
    ...stringCards(diffTipsFs(root, source, locale), LIST_TIPS_FS, label),
    ...stringCards(diffAbyssBuffs(root, source, locale), LIST_ABYSS, label),
    ...travelBuffCards(diffTravelBuffs(root, source, locale), label),
  ];
}

/** Every untranslated entry of `locale`, as Trello cards. */
export function collectCards(request: TrelloExportRequest): TrelloCard[] {
  const resolved = resolveRequest(request);
  return [...almanacListCards(resolved), ...stringListCards(resolved)];
}

export function exportTrello(options: TrelloExportOptions): TrelloExportResult {
  const resolved = resolveRequest(options);
  const outputDir = path.join(options.exportsRoot, resolved.locale);
  mkdirSync(outputDir, { recursive: true });

  const lists = writeTrelloCsvsByList(collectCards(options), outputDir);
  const readmePath = buildTrelloReadme({
    locale: resolved.locale,
    label: resolved.label,
    outputPath: path.join(outputDir, README_FILENAME),
    lists,
  });

  return {
    locale: resolved.locale,
    outputDir,
    readmePath,
    lists,
    totalCards: lists.reduce((sum, list) => sum + list.cardCount, 0),
  };
}
